server/src/main/java/io/druid/segment/realtime/RealtimeManager.java

```diff
@@ -319,6 +299,128 @@ public void run()
       }
     }
 
+    private void runFirehoseV2(FirehoseV2 firehose)
+    {
+      final Period intermediatePersistPeriod = config.getIntermediatePersistPeriod();
+      try {
+        firehose.start();
+      }
+      catch (Exception e) {
+      	log.error(e, "Failed to start firehoseV2");
+      	return;
+      }
+      long nextFlush = new DateTime().plus(intermediatePersistPeriod).getMillis();
+
+      boolean haveRow = true;
+      while (haveRow) {
+        InputRow inputRow = null;
+        int numRows = 0;
+        try {
+          inputRow = firehose.currRow();
+          if (inputRow != null) {
+            try {
+              numRows = plumber.add(inputRow);
+            }
+            catch (IndexSizeExceededException e) {
+              log.info("Index limit exceeded: %s", e.getMessage());
+              nextFlush = doIncrementalPersist(firehose.makeCommitter(), intermediatePersistPeriod);
+              continue;
+            }
+
+            if (numRows < 0) {
+              metrics.incrementThrownAway();
+              log.debug("Throwing away event[%s]", inputRow);
+            } else {
+              metrics.incrementProcessed();
+            }
+          }
+        }
+        catch (Exception e) {
+          log.makeAlert(e, "Some exception got thrown while processing rows.  Ignoring and continuing.")
```

Comment:
This is the same code as I copied from https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/segment/realtime/RealtimeManager.java#L241. I have to split this part out from `run()` because firehose v1 and v2's ingestion logic is slightly different. Firehose V1 goes like "while has more, parse, index", and firehose V2 first `start()` to ingest on row, then it goes like "if parse and ingest successful, then advance to next row, else keep ingesting this row"